### ISSUE TITLE
new Seurat 3.0.0 cell object indexing problem

### ISSUE THREAD

**Issue by krejciadam** [satijalab/seurat#1492]
Hi!
I've been using Seurat 3.0.0 for some time and yesterday (2nd May), I noticed it started to be the official version, so I installed it from CRAN. To my surprise, the scripts I've created with older Seurat 3.0.0 (installed on 30th January 2019 using devoolts) started to produce totally incorrect results. I've tracked the problem to re-ordering the cells in a Seurat object by matrix coordinates. This is an example: 

```
> data <- Read10X(data.dir = "/path/to/3k_pbmcs_example/filtered_gene_bc_matrices/hg19/")
> cells <- CreateSeuratObject(counts = data, project = "pbmc3k", min.cells = 3, min.features = 200)
Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')

#jutst 10 cells: 
> cells <- cells[, 1:10]

#to see in what order the cells are in the object:
> cells[['order']] <- c(1:10)
> cells$order
AAACATACAACCAC AAACATTGAGCTAC AAACATTGATCAGC AAACCGTGCTTCCG AAACCGTGTATGCG AAACGCACTGGTAC AAACGCTGACCAGT AAACGCTGGTTCTT AAACGCTGTAGCCA 
             1              2              3              4              5              6              7              8              9 
AAACGCTGTTTCTG 
            10 

#now subset the cells. This is OK: 
> cells[, 5:8]$order
AAACCGTGTATGCG AAACGCACTGGTAC AAACGCTGACCAGT AAACGCTGGTTCTT 
             5              6              7              8 

#this is also OK: 
> cells[, 8:5]$order
AAACGCTGGTTCTT AAACGCTGACCAGT AAACGCACTGGTAC AAACCGTGTATGCG 
             8              7              6              5 

#this is still OK: 
> cells[, 10:2]$order
AAACGCTGTTTCTG AAACGCTGTAGCCA AAACGCTGGTTCTT AAACGCTGACCAGT AAACGCACTGGTAC AAACCGTGTATGCG AAACCGTGCTTCCG AAACATTGATCAGC AAACATTGAGCTAC 
            10              9              8              7              6              5              4              3              2 

#selecting all cells in a different order is not OK. Order is not changed: 
> cells[, 10:1]$order
AAACATACAACCAC AAACATTGAGCTAC AAACATTGATCAGC AAACCGTGCTTCCG AAACCGTGTATGCG AAACGCACTGGTAC AAACGCTGACCAGT AAACGCTGGTTCTT AAACGCTGTAGCCA 
             1              2              3              4              5              6              7              8              9 
AAACGCTGTTTCTG 
            10 

#selecting all the cells in any order does not reorder them: 
> cells[, sample(c(1:10))]$order
AAACATACAACCAC AAACATTGAGCTAC AAACATTGATCAGC AAACCGTGCTTCCG AAACCGTGTATGCG AAACGCACTGGTAC AAACGCTGACCAGT AAACGCTGGTTCTT AAACGCTGTAGCCA 
             1              2              3              4              5              6              7              8              9 
AAACGCTGTTTCTG 
            10 
```

The thing that messes my scripts is that I'm re-ordering some cells and then assigning some labels to them. But cells are actually not re-ordered, so incorrect labels are then assigned to them: 

```
> cells_rev <- cells[, 10:1]
> cells_rev[['order_rev']] <- c(1:10)

#cells were actually not reversed:
> cells_rev[, 2]$order
AAACATTGAGCTAC 
             2 
#So they got incorrect reversed index: 
> cells_rev[, 2]$order_rev
AAACATTGAGCTAC 
             2 
```

The 30th January version does actually not work as expected either, but for the sake of assigning labels, it is good enough: 
```
> cells_rev <- cells[, 10:1]
> cells_rev[['order_rev']] <- c(1:10)

#now this is actually still weird, as the 2nd cell of the reversed object is not the 9th cell
#of the original object, but still the 2nd cell of the original object.
> cells_rev[, 2]$order
AAACATTGAGCTAC 
             2 

#Nevertheless, it got the correct reversed index
> cells_rev[, 2]$order_rev
AAACATTGAGCTAC 
             9 
```

I guess users can do without any use of such ordering, but I still consider this a pretty weird behavior. Thank you.
Adam

**Comment by mojaveazure** [comment]
Hi Adam,

Neither `subset` nor `[` in Seurat v3 are designed for reordering cells. If all cells are passed to either one of these functions, it currently just returns the object as is. I'm closing this issue for now, but if you're up for it, we would love a PR adding this functionality to `subset.Seurat` (the underlying function for `[.Seurat`

**Comment by davidlieb** [comment]
I just discovered this issue from a couple years ago where trying to randomly sort all the cells in a Seurat object with "[ " returns the object unchanged. I'm using Seurat v4.04 and R v4.1.1. (I worked around the issue by merging random orderings of the first and second halves of the object.) It is quite an unexpected result. Is there any change in status? BTW, what does PR mean? 

Thanks for Seurat!

David
